In [7]:
#!pip install scikit-learn
#!pip install seaborn
#!pip install matplotlib

In [8]:
# importamos las librerías que necesitamos

# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
import re

# Imputación de nulos usando métodos avanzados estadísticos
# -----------------------------------------------------------------------
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer

# Librerías de visualización
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt
# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# Punto --> Exploración de los datos! ✅

In [9]:
df = pd.read_csv('../HR RAW DATA.csv', index_col= 0)
df.head()

Age Attrition BusinessTravel DailyRate                Department  \
0  51        No            NaN    684,0$                       NaN   
1  52        No            NaN    699,0$                       NaN   
2  42        No  travel_rarely    532,0$   Research & Development    
3  47        No  travel_rarely    359,0$                       NaN   
4  46        No            NaN   1319,0$                       NaN   

   DistanceFromHome  Education    EducationField  employeecount  \
0                 6          3               NaN              1   
1                 1          4     Life Sciences              1   
2                 4          2  Technical Degree              1   
3                 2          4           Medical              1   
4                 3          3  Technical Degree              1   

  employeenumber  EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  \
0          162,0                        1       0         51               3   
1          259,0                        3       0         65               2   
2          319,0                        3       0         58               3   
3            NaN                        1       1         82               3   
4            NaN                        1       1         45               4   

   JobLevel              JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0         5   resEArch DIREcToR                 3           NaN       19537,0   
1         5             ManAGeR                 3           NaN       19999,0   
2         5             ManaGER                 4       Married       19232,0   
3         4   ReseArCH DIrECtOr                 3       Married       17169,0   
4         4     sAleS EXECUtIve                 1      Divorced           NaN   

   MonthlyRate  NUMCOMPANIESWORKED Over18 OverTime  PercentSalaryHike  \
0         6462                   7      Y       No                 13   
1         5678                   0    NaN      NaN                 14   
2         4933                   1    NaN       No                 11   
3        26703                   3      Y      NaN                 19   
4         7739                   2      Y       No                 12   

  PerformanceRating  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0               3,0                         3           NaN                 0   
1               3,0                         1           NaN                 1   
2               3,0                         4           NaN                 0   
3               3,0                         2           NaN                 2   
4               3,0                         4           NaN                 1   

  TOTALWORKINGYEARS  TrainingTimesLastYear WORKLIFEBALANCE  YearsAtCompany  \
0               NaN                      5             3,0              20   
1              34,0                      5             3,0              33   
2              22,0                      3             NaN              22   
3               NaN                      2             NaN              20   
4               NaN                      5             3,0              19   

  YearsInCurrentRole  YearsSinceLastPromotion  YEARSWITHCURRMANAGER  \
0                NaN                       15                    15   
1                NaN                       11                     9   
2                NaN                       11                    15   
3                NaN                        5                     6   
4                NaN                        2                     8   

  SameAsMonthlyIncome  DateBirth       Salary  \
0             19537,0       1972  1000000000$   
1             19999,0       1971  1000000000$   
2             19232,0       1981  1000000000$   
3             17169,0       1976  1000000000$   
4                 NaN       1977  1000000000$   

                        RoleDepartament  NUMBERCHILDREN RemoteWork  
0                       

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1614 entries, 0 to 1613
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1614 non-null   object 
 1   Attrition                 1614 non-null   object 
 2   BusinessTravel            842 non-null    object 
 3   DailyRate                 1614 non-null   object 
 4   Department                302 non-null    object 
 5   DistanceFromHome          1614 non-null   int64  
 6   Education                 1614 non-null   int64  
 7   EducationField            869 non-null    object 
 8   employeecount             1614 non-null   int64  
 9   employeenumber            1183 non-null   object 
 10  EnvironmentSatisfaction   1614 non-null   int64  
 11  Gender                    1614 non-null   int64  
 12  HourlyRate                1614 non-null   object 
 13  JobInvolvement            1614 non-null   int64  
 14  JobLevel     

In [11]:
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'employeecount',
       'employeenumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NUMCOMPANIESWORKED',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TOTALWORKINGYEARS', 'TrainingTimesLastYear', 'WORKLIFEBALANCE',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YEARSWITHCURRMANAGER', 'SameAsMonthlyIncome', 'DateBirth', 'Salary',
       'RoleDepartament', 'NUMBERCHILDREN', 'RemoteWork'],
      dtype='object')

In [12]:
df.dtypes

Age                          object
Attrition                    object
BusinessTravel               object
DailyRate                    object
Department                   object
DistanceFromHome              int64
Education                     int64
EducationField               object
employeecount                 int64
employeenumber               object
EnvironmentSatisfaction       int64
Gender                        int64
HourlyRate                   object
JobInvolvement                int64
JobLevel                      int64
JobRole                      object
JobSatisfaction               int64
MaritalStatus                object
MonthlyIncome                object
MonthlyRate                   int64
NUMCOMPANIESWORKED            int64
Over18                       object
OverTime                     object
PercentSalaryHike             int64
PerformanceRating            object
RelationshipSatisfaction      int64
StandardHours                object
StockOptionLevel            

In [13]:
# Valores nulos del total dataframe por columnas
df.isnull().sum()

Age                            0
Attrition                      0
BusinessTravel               772
DailyRate                      0
Department                  1312
DistanceFromHome               0
Education                      0
EducationField               745
employeecount                  0
employeenumber               431
EnvironmentSatisfaction        0
Gender                         0
HourlyRate                     0
JobInvolvement                 0
JobLevel                       0
JobRole                        0
JobSatisfaction                0
MaritalStatus                651
MonthlyIncome                843
MonthlyRate                    0
NUMCOMPANIESWORKED             0
Over18                       901
OverTime                     676
PercentSalaryHike              0
PerformanceRating            195
RelationshipSatisfaction       0
StandardHours               1195
StockOptionLevel               0
TOTALWORKINGYEARS            526
TrainingTimesLastYear          0
WORKLIFEBA

In [14]:
porcentaje_nulos = df.isnull().mean() * 100

# df.isnull().sum()/registro[0] - Otra forma de hacerlo

# Muestra los resultados<z
porcentaje_nulos

Age                           0.000000
Attrition                     0.000000
BusinessTravel               47.831475
DailyRate                     0.000000
Department                   81.288724
DistanceFromHome              0.000000
Education                     0.000000
EducationField               46.158612
employeecount                 0.000000
employeenumber               26.703841
EnvironmentSatisfaction       0.000000
Gender                        0.000000
HourlyRate                    0.000000
JobInvolvement                0.000000
JobLevel                      0.000000
JobRole                       0.000000
JobSatisfaction               0.000000
MaritalStatus                40.334572
MonthlyIncome                52.230483
MonthlyRate                   0.000000
NUMCOMPANIESWORKED            0.000000
Over18                       55.824040
OverTime                     41.883519
PercentSalaryHike             0.000000
PerformanceRating            12.081784
RelationshipSatisfaction 

In [15]:
registro = df.shape
registro

(1614, 41)

# Punto --> Revisión de cada columna 👌🏼

In [16]:
 
'Age' # OK
'Attrition' # OK
'BusinessTravel' # revisar los NAN
'DailyRate' # OK
'Department' # ELIMINDADO
'DistanceFromHome' # OK
'Education' # OK
'EducationField' # Revisar los NAN
'employeecount' # Ok - ELIMINDADO
'employeenumber' # duplicados

'EnvironmentSatisfaction' # Ok 
'Gender' # # OK
'HourlyRate' # OK
'JobInvolvement' # OK
'JobLevel' # OK
'JobRole' # OK
'JobSatisfaction' # OK
'MaritalStatus' # revisar los NAN
'MonthlyIncome' # revisar los NAN
'MonthlyRate' # OK

'NUMCOMPANIESWORKED' # OK
'Over18' # ELIMINADO
'OverTime' # revisar los NAN
'PercentSalaryHike' # OK
'PerformanceRating' # revisar los NAN
'RelationshipSatisfaction' # OK
'StandardHours' # ELIMINADO
'StockOptionLevel' # OK
'TOTALWORKINGYEARS' # revisar los NAN
'TrainingTimesLastYear' # OK

'WORKLIFEBALANCE' # revisar los NAN
'YearsAtCompany' # OK
'YearsInCurrentRole' # ELIMINADO
'YearsSinceLastPromotion' # OK
'YEARSWITHCURRMANAGER' # OK
'SameAsMonthlyIncome' # ELIMINADO
'DateBirth' # OK
'Salary' # ELIMINADO
'RoleDepartament' # ELIMINADO
'NUMBERCHILDREN' # ELIMINADO
'RemoteWork' # OK

'RemoteWork'

# Eliminar columnas ✅

- `employeecount` # Porque siempre es el mismo valor y es un contador de los empleados
- `Department` # con un 81% de datos nulos y ya una columna en paralelo dice lo mismos datos
- `Over18` # Porque ya tenemos la edad y tiene muchos valores perdidos
- `StandardHours` # Porque todos tienen el mismo valor que es 80 y ademas tienen muchos valores perdidos
- `Salary` # Porque es igual a 10000000$ en todos los valores del dataframe
- `NUMBERCHILDREN` # Porque no tiene ningun valor y es todo con valores missing.
- `roledepartament` # con un 81% de datos nulos y ya una columna en paralelo dice lo mismos datos
- `YearsInCurrentRole` # 97% de nulos 
- `SameAsMonthlyIncome` # Contiene la misma informacion que MonthlyIncome, por lo tanto la eliminamos


In [17]:
df.drop(["employeecount", "Department", "Over18", "StandardHours", "YearsInCurrentRole", "RoleDepartament", "Salary", "NUMBERCHILDREN"], axis=1, inplace=True )

In [18]:
df.head(1)

Age Attrition BusinessTravel DailyRate  DistanceFromHome  Education  \
0  51        No            NaN    684,0$                 6          3   

  EducationField employeenumber  EnvironmentSatisfaction  Gender HourlyRate  \
0            NaN          162,0                        1       0         51   

   JobInvolvement  JobLevel              JobRole  JobSatisfaction  \
0               3         5   resEArch DIREcToR                 3   

  MaritalStatus MonthlyIncome  MonthlyRate  NUMCOMPANIESWORKED OverTime  \
0           NaN       19537,0         6462                   7       No   

   PercentSalaryHike PerformanceRating  RelationshipSatisfaction  \
0                 13               3,0                         3   

   StockOptionLevel TOTALWORKINGYEARS  TrainingTimesLastYear WORKLIFEBALANCE  \
0                 0               NaN                      5             3,0   

   YearsAtCompany  YearsSinceLastPromotion  YEARSWITHCURRMANAGER  \
0              20                       15                    15   

  SameAsMonthlyIncome  DateBirth RemoteWork  
0             19537,0       1972        Yes

In [19]:
df.shape

(1614, 33)

----------

- `Homogeneizamos el nombre de las columnas a minúsculas` ✅

In [20]:
# vamos a iterar por todas las columnas del DataFrame y a cada una de ellas la pondremos en minúsula y le aplicaremos un replace para quitar los puntos. 
nuevas_columnas = {columna: columna.lower() for columna in df.columns}

# comprobamos que hemos creado el diccionario correctamente
nuevas_columnas

{'Age': 'age',
 'Attrition': 'attrition',
 'BusinessTravel': 'businesstravel',
 'DailyRate': 'dailyrate',
 'DistanceFromHome': 'distancefromhome',
 'Education': 'education',
 'EducationField': 'educationfield',
 'employeenumber': 'employeenumber',
 'EnvironmentSatisfaction': 'environmentsatisfaction',
 'Gender': 'gender',
 'HourlyRate': 'hourlyrate',
 'JobInvolvement': 'jobinvolvement',
 'JobLevel': 'joblevel',
 'JobRole': 'jobrole',
 'JobSatisfaction': 'jobsatisfaction',
 'MaritalStatus': 'maritalstatus',
 'MonthlyIncome': 'monthlyincome',
 'MonthlyRate': 'monthlyrate',
 'NUMCOMPANIESWORKED': 'numcompaniesworked',
 'OverTime': 'overtime',
 'PercentSalaryHike': 'percentsalaryhike',
 'PerformanceRating': 'performancerating',
 'RelationshipSatisfaction': 'relationshipsatisfaction',
 'StockOptionLevel': 'stockoptionlevel',
 'TOTALWORKINGYEARS': 'totalworkingyears',
 'TrainingTimesLastYear': 'trainingtimeslastyear',
 'WORKLIFEBALANCE': 'worklifebalance',
 'YearsAtCompany': 'yearsatcompany'

In [21]:
# una vez que tenemos el diccionario vamos a aplicar el método rename
df.rename(columns = nuevas_columnas, inplace = True)

# comprobamos que se han cambiado los nombres de las columnas
df.head()

age attrition businesstravel dailyrate  distancefromhome  education  \
0  51        No            NaN    684,0$                 6          3   
1  52        No            NaN    699,0$                 1          4   
2  42        No  travel_rarely    532,0$                 4          2   
3  47        No  travel_rarely    359,0$                 2          4   
4  46        No            NaN   1319,0$                 3          3   

     educationfield employeenumber  environmentsatisfaction  gender  \
0               NaN          162,0                        1       0   
1     Life Sciences          259,0                        3       0   
2  Technical Degree          319,0                        3       0   
3           Medical            NaN                        1       1   
4  Technical Degree            NaN                        1       1   

  hourlyrate  jobinvolvement  joblevel              jobrole  jobsatisfaction  \
0         51               3         5   resEArch DIREcToR                 3   
1         65               2         5             ManAGeR                 3   
2         58               3         5             ManaGER                 4   
3         82               3         4   ReseArCH DIrECtOr                 3   
4         45               4         4     sAleS EXECUtIve                 1   

  maritalstatus monthlyincome  monthlyrate  numcompaniesworked overtime  \
0           NaN       19537,0         6462                   7       No   
1           NaN       19999,0         5678                   0      NaN   
2       Married       19232,0         4933                   1       No   
3       Married       17169,0        26703                   3      NaN   
4      Divorced           NaN         7739                   2       No   

   percentsalaryhike performancerating  relationshipsatisfaction  \
0                 13               3,0                         3   
1                 14               3,0                         1   
2                 11               3,0                         4   
3                 19               3,0                         2   
4                 12               3,0                         4   

   stockoptionlevel totalworkingyears  trainingtimeslastyear worklifebalance  \
0                 0               NaN                      5             3,0   
1                 1              34,0                      5             3,0   
2                 0              22,0                      3             NaN   
3                 2               NaN                      2             NaN   
4                 1               NaN                      5             3,0   

   yearsatcompany  yearssincelastpromotion  yearswithcurrmanager  \
0              20                       15                    15   
1              33                       11                     9   
2              22                       11                    15   
3              20                        5                     6   
4              19                        2                     8   

  sameasmonthlyincome  datebirth remotework  
0             19537,0       1972        Yes  
1             19999,0       1971          1  
2             19232,0       1981          1  
3             17169,0       1976      False  
4                 NaN       1977          0

--------

- `Cambiamos las letras de números en la columna (age)` ✅
- `colocamos los datos de object a int` ✅
- `Falta guerdarlo directamente en el dataframe` ✅

In [22]:
# Diccionario para convertir palabras a números
word_to_number = {
    'zero': 0, 'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5,
    'six': 6, 'seven': 7, 'eight': 8, 'nine': 9, 'ten': 10, 'eleven': 11,
    'twelve': 12, 'thirteen': 13, 'fourteen': 14, 'fifteen': 15,
    'sixteen': 16, 'seventeen': 17, 'eighteen': 18, 'nineteen': 19,
    'twenty': 20, 'thirty': 30, 'forty': 40, 'fifty': 50, 'sixty': 60,
    'seventy': 70, 'eighty': 80, 'ninety': 90
}

# Función para convertir cadenas a números
def convert_to_number(age):
    try:
        # Primero intentamos convertir directamente a entero
        return int(age)
    except ValueError:
        # Si falla, intentamos convertir usando el diccionario
        words = age.split('-')
        number = 0
        for word in words:
            if word in word_to_number:
                number += word_to_number[word]
            else:
                return f"Error converting '{age}'"
        return number

# Aplicar la función de conversión a la columna 'Age' del DataFrame
df['age'] = df['age'].apply(convert_to_number)

# Verificar si hay errores en la conversión
errors = df['age'][df['age'].apply(lambda x: isinstance(x, str))]

# Mostrar los resultados
print("Converted Ages in DataFrame:")
print(df['age'].head())

if not errors.empty:
    print("\nErrors:")
    print(errors)
else:
    print("\nNo conversion errors.")


Converted Ages in DataFrame:
0    51
1    52
2    42
3    47
4    46
Name: age, dtype: int64

No conversion errors.


In [23]:
df['age'].unique()

array([51, 52, 42, 47, 46, 48, 59, 41, 56, 38, 55, 40, 58, 35, 45, 33, 36,
       34, 53, 43, 60, 32, 37, 49, 39, 50, 44, 30, 29, 31, 54, 57, 27, 28,
       26, 25, 24, 23, 22, 21, 20, 19, 18])

In [24]:
df['age'].shape

(1614,)

---------------

No requiere cambios ✅

In [25]:
df['attrition'].unique()

array(['No', 'Yes'], dtype=object)

---------

- `Separamos los string con espacio vacio('') y quitamos el (_)`✅
- `Eliminar los NAN`❓

In [26]:
df['businesstravel'].unique() 

array([nan, 'travel_rarely', 'travel_frequently', 'non-travel'],
      dtype=object)

In [27]:
df["businesstravel"] = df["businesstravel"].str.replace("_", " ").str.replace("-"," ")
df.sample(5)

age attrition     businesstravel dailyrate  distancefromhome  education  \
300    35        No                NaN   1315,0$                22          3   
101    43        No      travel rarely   1291,0$                15          2   
763    37        No  travel frequently   1278,0$                 1          4   
1033   40        No      travel rarely    329,0$                 1          4   
1372   32        No                NaN    862,0$                 2          1   

     educationfield employeenumber  environmentsatisfaction  gender  \
300   Life Sciences          381,0                        2       1   
101             NaN         1666,0                        3       0   
763             NaN         1700,0                        3       0   
1033  Life Sciences         1361,0                        2       0   
1372            NaN         1190,0                        3       1   

     hourlyrate  jobinvolvement  joblevel                  jobrole  \
300          71               4         3                 maNagER    
101          65               2         4       ReSearch DiRecTOR    
763          31               1         2         SaLes EXECUTivE    
1033         88               3         1   LaBOratoRy TECHNICiAn    
1372         76               3         1    SAlES rePrESentaTIVe    

      jobsatisfaction maritalstatus monthlyincome  monthlyrate  \
300                 2      Divorced           NaN        19100   
101                 3           NaN           NaN         3525   
763                 4           NaN           NaN         7677   
1033                2           NaN           NaN         6762   
1372                1           NaN        2827,0        14947   

      numcompaniesworked overtime  percentsalaryhike performancerating  \
300                    7      NaN                 18               NaN   
101                    1       No                 24               4,0   
763                    1      NaN                 14               3,0   
1033                   3       No                 22               4,0   
1372                   1      NaN                 12               3,0   

      relationshipsatisfaction  stockoptionlevel totalworkingyears  \
300                          2                 1              10,0   
101                          1                 1               NaN   
763                          3                 2               6,0   
1033                         3                 1               7,0   
1372                         3                 3               1,0   

      trainingtimeslastyear worklifebalance  yearsatcompany  \
300                       6             2,0               7   
101                       3             3,0              14   
763                       2             2,0               6   
1033                      3             3,0               4   
1372                      3             3,0               1   

      yearssincelastpromotion  yearswithcurrmanager sameasmonthlyincome  \
300                         6                     2                 NaN   
101                         6                    11                 NaN   
763                         1                     3                 NaN   
1033                        0                     3                 NaN   
1372                        0                     0              2827,0   

      datebirth remotework  
300        1988          0  
101        1980          1  
763        1986          1  
1033       1983        Yes  
1372       1991       True

--------

- `Cambiar a float` ✅
- `Cambiar las comas(,) por puntos(.)` ✅

In [28]:
df['dailyrate'] = df['dailyrate'].str.replace(",", ".").str.replace("$", "")
df['dailyrate'].unique()

array(['684.0', '699.0', '532.0', '359.0', '1319.0', '117.0', '1435.0',
       '635.0', '1276.0', '840.0', '247.0', '1369.0', '201.0', '1360.0',
       '692.0', '1398.0', '286.0', '1402.0', '819.0', '884.0', '1238.0',
       '515.0', '1223.0', '202.0', '928.0', '607.0', '266.0', '429.0',
       '589.0', 'nan', '1180.0', '1282.0', '776.0', '665.0', '526.0',
       '1034.0', '1403.0', '1499.0', '580.0', '859.0', '263.0', '1376.0',
       '885.0', '1003.0', '1321.0', '394.0', '1372.0', '1333.0', '228.0',
       '737.0', '823.0', '667.0', '301.0', '573.0', '1329.0', '630.0',
       '1063.0', '1017.0', '1296.0', '939.0', '1355.0', '1448.0', '200.0',
       '1202.0', '404.0', '208.0', '813.0', '465.0', '1189.0', '1001.0',
       '1394.0', '161.0', '288.0', '682.0', '1354.0', '147.0', '119.0',
       '1413.0', '452.0', '334.0', '1132.0', '982.0', '480.0', '1099.0',
       '672.0', '1379.0', '583.0', '1492.0', '1050.0', '469.0', '237.0',
       '1440.0', '1291.0', '1157.0', '1336.0', '1224.0',

In [29]:
df['dailyrate'] = df['dailyrate'].astype(float)

In [30]:
df['dailyrate'].unique()

array([ 684.,  699.,  532.,  359., 1319.,  117., 1435.,  635., 1276.,
        840.,  247., 1369.,  201., 1360.,  692., 1398.,  286., 1402.,
        819.,  884., 1238.,  515., 1223.,  202.,  928.,  607.,  266.,
        429.,  589.,   nan, 1180., 1282.,  776.,  665.,  526., 1034.,
       1403., 1499.,  580.,  859.,  263., 1376.,  885., 1003., 1321.,
        394., 1372., 1333.,  228.,  737.,  823.,  667.,  301.,  573.,
       1329.,  630., 1063., 1017., 1296.,  939., 1355., 1448.,  200.,
       1202.,  404.,  208.,  813.,  465., 1189., 1001., 1394.,  161.,
        288.,  682., 1354.,  147.,  119., 1413.,  452.,  334., 1132.,
        982.,  480., 1099.,  672., 1379.,  583., 1492., 1050.,  469.,
        237., 1440., 1291., 1157., 1336., 1224.,  735., 1389.,  638.,
       1240.,  194., 1339.,  111., 1469.,  470., 1232., 1249.,  798.,
        549.,  570.,  541.,  164., 1117.,  619.,  319.,  956., 1245.,
       1397.,  527.,  213.,  882.,  330.,  406.,  217.,  481.,  669.,
       1465.,  685.,

In [31]:
df.sample()

age attrition businesstravel  dailyrate  distancefromhome  education  \
1480   43        No            NaN      982.0                12          3   

     educationfield employeenumber  environmentsatisfaction  gender  \
1480            NaN            NaN                        1       0   

         hourlyrate  jobinvolvement  joblevel              jobrole  \
1480  Not Available               2         4   ReseArCH dirECTOR    

      jobsatisfaction maritalstatus monthlyincome  monthlyrate  \
1480                2           NaN           NaN         4345   

      numcompaniesworked overtime  percentsalaryhike performancerating  \
1480                   1       No                 11               3,0   

      relationshipsatisfaction  stockoptionlevel totalworkingyears  \
1480                         3                 1               NaN   

      trainingtimeslastyear worklifebalance  yearsatcompany  \
1480                      3             NaN              25   

      yearssincelastpromotion  yearswithcurrmanager sameasmonthlyincome  \
1480                        3                     9                 NaN   

      datebirth remotework  
1480       1980          1

--------

Eliminado ❌

In [32]:
df['department'].unique()

KeyError: 'department'

In [ ]:
df['department'].isnull().sum() # Tiene un 81% de nulos

1312

-----------

`Cambiar los números a positivos los que están negativos`✅

In [ ]:
df['distancefromhome'] = df['distancefromhome'].abs()

In [ ]:
df['distancefromhome'].unique()


array([ 6,  1,  4,  2,  3, 22, 25,  9,  7, 23, 10, 12, 14, 13, 15,  8, 42,
       28, 37,  5, 16, 35, 26, 24, 29, 17, 21, 18, 30, 27, 20, 31, 39, 11,
       19, 33, 34, 46, 36, 45, 47, 32, 41, 49, 48, 38, 43, 40, 44])

---------

No requiere cambios ✅

In [ ]:
df['education'].unique()

array([3, 4, 2, 1, 5])

--------

- `cambiar a minúscula todo` ✅
- `Eliminar los NAN`

In [ ]:
df['educationfield'] = df['educationfield'].str.lower()
df.head()

age attrition businesstravel  dailyrate  distancefromhome  education  \
0   51        No            NaN      684.0                 6          3   
1   52        No            NaN      699.0                 1          4   
2   42        No  travel rarely      532.0                 4          2   
3   47        No  travel rarely      359.0                 2          4   
4   46        No            NaN     1319.0                 3          3   

     educationfield employeenumber  environmentsatisfaction  gender  \
0               NaN          162,0                        1       0   
1     life sciences          259,0                        3       0   
2  technical degree          319,0                        3       0   
3           medical            NaN                        1       1   
4  technical degree            NaN                        1       1   

  hourlyrate  jobinvolvement  joblevel              jobrole  jobsatisfaction  \
0         51               3         5   resEArch DIREcToR                 3   
1         65               2         5             ManAGeR                 3   
2         58               3         5             ManaGER                 4   
3         82               3         4   ReseArCH DIrECtOr                 3   
4         45               4         4     sAleS EXECUtIve                 1   

  maritalstatus monthlyincome  monthlyrate  numcompaniesworked overtime  \
0           NaN       19537,0         6462                   7       No   
1           NaN       19999,0         5678                   0      NaN   
2       Married       19232,0         4933                   1       No   
3       Married       17169,0        26703                   3      NaN   
4      Divorced           NaN         7739                   2       No   

   percentsalaryhike performancerating  relationshipsatisfaction  \
0                 13               3,0                         3   
1                 14               3,0                         1   
2                 11               3,0                         4   
3                 19               3,0                         2   
4                 12               3,0                         4   

   stockoptionlevel totalworkingyears  trainingtimeslastyear worklifebalance  \
0                 0               NaN                      5             3,0   
1                 1              34,0                      5             3,0   
2                 0              22,0                      3             NaN   
3                 2               NaN                      2             NaN   
4                 1               NaN                      5             3,0   

   yearsatcompany  yearssincelastpromotion  yearswithcurrmanager  \
0              20                       15                    15   
1              33                       11                     9   
2              22                       11                    15   
3              20                        5                     6   
4              19                        2                     8   

  sameasmonthlyincome  datebirth remotework  
0             19537,0       1972        Yes  
1             19999,0       1971          1  
2             19232,0       1981          1  
3             17169,0       1976      False  
4                 NaN       1977          0

In [ ]:
df['educationfield'].unique()

array([nan, 'life sciences', 'technical degree', 'medical', 'other',
       'marketing', 'human resources'], dtype=object)

--------

Eliminado ❌

In [ ]:
df['employeecount'].unique()

-------

# `Eliminar duplicados`
- `Revisar bien cada registro para tomar una decisión sobre los nulos que tiene filas llenas pero el id es nulo`

In [ ]:
df['employeenumber'].unique()

array(['162,0', '259,0', '319,0', ..., '2012,0', '2023,0', '2040,0'],
      dtype=object)

In [ ]:
df['employeenumber'].duplicated().sum()

534

In [ ]:
df['employeenumber'].value_counts()

employeenumber
482,0     2
530,0     2
507,0     2
517,0     2
522,0     2
         ..
161,0     1
164,0     1
190,0     1
194,0     1
2040,0    1
Name: count, Length: 1079, dtype: int64

In [ ]:
df.duplicated(subset=['employeenumber'])

0       False
1       False
2       False
3       False
4        True
        ...  
1609     True
1610     True
1611     True
1612     True
1613     True
Length: 1614, dtype: bool

In [ ]:
df_duplicated = df.loc[df.employeenumber == df.employeenumber, :]
df_duplicated.sample(10)

employeenumber  age attrition     businesstravel  dailyrate  \
577           355,0   31       Yes                NaN      307.0   
1389         1306,0   45        No  travel frequently      364.0   
701           599,0   41        No         non travel      267.0   
1050         1436,0   36        No                NaN      329.0   
1078         1592,0   23        No                NaN      977.0   
1393         1427,0   31       Yes  travel frequently        NaN   
796          2068,0   34        No                NaN        NaN   
227          1356,0   34        No                NaN      507.0   
504          1085,0   41        No                NaN      509.0   
1265          309,0   59        No      travel rarely      142.0   

      distancefromhome  education educationfield  environmentsatisfaction  \
577                 29          2            NaN                        3   
1389                25          3        medical                        2   
701                 10          2            NaN                        4   
1050                16          4      marketing                        3   
1078                10          3            NaN                        4   
1393                 1          4  life sciences                        2   
796                  8          3            NaN                        2   
227                 15          2        medical                        3   
504                  7          2            NaN                        2   
1265                 3          3  life sciences                        3   

      gender     hourlyrate  jobinvolvement  joblevel  \
577        0             71               2         1   
1389       1  Not Available               3         5   
701        0  Not Available               3         2   
1050       1             98               2         2   
1078       0             45               4         1   
1393       1             50               1         1   
796        0             82               4         2   
227        1             66               3         2   
504        1             43               4         1   
1265       0             70               2         1   

                      jobrole  jobsatisfaction maritalstatus monthlyincome  \
577    laboraTORy teChnICiaN                 2           NaN           NaN   
1389                 manAGEr                 2           NaN       18061,0   
701          SAlES ExecutIve                 4        Single           NaN   
1050         SaLEs eXEcUtiVe                 1           NaN        5647,0   
1078      rESEarcH SCIEnTist                 3       Married           NaN   
1393    sALeS RePReSENtAtiVe                 3        Single           NaN   
796    lABoRaToRy TEChnicIAN                 3           NaN        4404,0   
227          SALEs eXeCuTive                 1        Single           NaN   
504       reseaRcH SCIEnTisT                 3           NaN        3376,0   
1265      ResEaRCh sciENtIST                 4       Married        2177,0   

      monthlyrate  numcompaniesworked overtime  percentsalaryhike  \
577         11652                   0       No                 11   
1389        13035                   3      NaN                 22   
701         13430                   7       No                 14   
1050        13494                   4       No                 13   
1078        12826                   2       No                 16   
1393        16154                   1      NaN                 12   
796         10228                   2       No                 12   
227         23553                   1      NaN                 12   
504         18863                   1       No                 13   
1265         8456                   3      NaN                 17   

     performancerating  relationshipsatisfaction  stockoptionlevel  \
577                3,0                         2                 0   
1389               4,

#### Muestra los duplicados juntos, se puede comparar cada fila

In [ ]:
# Identificar los números de 'employeenumber' que están duplicados
employeenumber_counts = df['employeenumber'].value_counts()
duplicated_employeenumbers = employeenumber_counts[employeenumber_counts > 1].index

# Filtrar el DataFrame original para mostrar solo las filas con 'employeenumber' duplicados
df_duplicated_employeenumbers = df[df['employeenumber'].isin(duplicated_employeenumbers)]

# Ordenar por 'employeenumber' para verlos agrupados
df_duplicated_employeenumbers_sorted = df_duplicated_employeenumbers.sort_values(by='employeenumber')

# Mostrar el DataFrame resultante
df_duplicated_employeenumbers_sorted


employeenumber  age attrition businesstravel  dailyrate  \
987          1044,0   33        No     non travel     1038.0   
1483         1044,0   33        No     non travel     1038.0   
1484         1053,0   26       Yes            NaN      342.0   
988          1053,0   26       Yes            NaN      342.0   
990          1069,0   46        No            NaN      430.0   
...             ...  ...       ...            ...        ...   
1609          967,0   36       Yes  travel rarely      530.0   
367           972,0   45        No     non travel      805.0   
1610          972,0   45        No     non travel      805.0   
369           990,0   36        No     non travel     1229.0   
1612          990,0   36        No     non travel     1229.0   

      distancefromhome  education    educationfield  environmentsatisfaction  \
987                  8          1     life sciences                        2   
1483                 8          1     life sciences                        2   
1484                 2          3     life sciences                        1   
988                  2          3     life sciences                        1   
990                 47          4           medical                       47   
...                ...        ...               ...                      ...   
1609                 3          1     life sciences                        3   
367                 47          2               NaN                       47   
1610                 4          2               NaN                        3   
369                  8          4  technical degree                        1   
1612                 8          4  technical degree                        1   

      gender     hourlyrate  jobinvolvement  joblevel  \
987        1  Not Available               2         1   
1483       1             88               2         1   
1484       0             57               3         1   
988        0             57               3         1   
990        0             40               3         5   
...      ...            ...             ...       ...   
1609       0             51               2         3   
367        0             57               3         2   
1610       0             57               3         2   
369        0             84               3         2   
1612       0             84               3         2   

                      jobrole  jobsatisfaction maritalstatus monthlyincome  \
987     sALES ReprEsentatiVE                 4           NaN           NaN   
1483    sales repResENTaTiVE                 4           NaN           NaN   
1484      reSEArcH SCiEnTIst                 1       Married        2042,0   
988       rEsEaRCH SCIenTiSt                 1       Married        2042,0   
990        RESeArCH dIrEcTOR                 4           NaN           NaN   
...                       ...              ...           ...           ...   
1609         saLEs ExeCUTiVe                 4       Married       10325,0   
367    lABORaTory tecHNiCiAN                 2           NaN        4447,0   
1610   LAboRaTOry tECHNiCIAn                 2           NaN        4447,0   
369          SaLes EXeCuTiVE                 4      Divorced           NaN   
1612         SaLes ExecUtIVe                 4      Divorced           NaN   

      monthlyrate  numcompaniesworked overtime  percentsalaryhike  \
987         21437                   0      NaN                 19   
1483        21437                   0      NaN                 19   
1484        15346                   6      NaN                 14   
988         15346                   6      NaN                 14   
990         21445                   9       No                 17   
...           ...                 ...      ...                ...   
1609         5518                   1      NaN                 11   
367         23163                   1      NaN                 12   
1610        23163               

### compara cada duplicado con sus columnas y ver similitud de info

In [ ]:
# Función para comparar filas duplicadas
def compare_duplicates(df, key_column):
    # Identificar los números de 'key_column' que están duplicados
    key_counts = df[key_column].value_counts()
    duplicated_keys = key_counts[key_counts > 1].index

    # Filtrar el DataFrame original para mostrar solo las filas con 'key_column' duplicados
    df_duplicated_keys = df[df[key_column].isin(duplicated_keys)]

    # Ordenar por 'key_column' para verlos agrupados
    df_duplicated_keys_sorted = df_duplicated_keys.sort_values(by=key_column)

    # Comparar filas duplicadas
    comparison_results = {}
    for key in duplicated_keys:
        duplicated_rows = df_duplicated_keys_sorted[df_duplicated_keys_sorted[key_column] == key]
        comparison = duplicated_rows.apply(lambda x: x.nunique() == 1)
        comparison_results[key] = comparison[comparison == False].index.tolist()

    return comparison_results

# Comparar filas duplicadas para la columna 'employeenumber'
comparison_results = compare_duplicates(df, 'employeenumber')

# Mostrar los resultados de la comparación
for key, differences in comparison_results.items():
    if differences:
        print(f"EmpleyeeNumber {key} tiene diferencias en las columnas: {differences}")
    else:
        print(f"EmpleyeeNumber {key} no tiene diferencias en las columnas")


EmpleyeeNumber 482,0 tiene diferencias en las columnas: ['jobrole', 'maritalstatus', 'totalworkingyears', 'remotework']
EmpleyeeNumber 530,0 tiene diferencias en las columnas: ['educationfield', 'jobrole', 'worklifebalance', 'remotework']
EmpleyeeNumber 507,0 tiene diferencias en las columnas: ['businesstravel', 'distancefromhome', 'jobrole', 'maritalstatus', 'monthlyincome', 'totalworkingyears', 'sameasmonthlyincome', 'remotework']
EmpleyeeNumber 517,0 tiene diferencias en las columnas: ['businesstravel', 'educationfield', 'hourlyrate', 'jobrole', 'maritalstatus', 'monthlyincome', 'sameasmonthlyincome', 'remotework']
EmpleyeeNumber 522,0 tiene diferencias en las columnas: ['businesstravel', 'jobrole', 'overtime', 'remotework']
EmpleyeeNumber 523,0 tiene diferencias en las columnas: ['businesstravel', 'educationfield', 'jobrole', 'totalworkingyears']
EmpleyeeNumber 525,0 tiene diferencias en las columnas: ['businesstravel', 'educationfield', 'jobrole', 'maritalstatus', 'overtime', 'tot

# Eliminar duplicados de `employeenumber`, limpiar columna y ver que hacer con los nulos

In [38]:
# Eliminar duplicados en la columna 'employeenumber' y conservar solo la primera ocurrencia
df.drop_duplicates(subset='employeenumber', keep='first', inplace = True)

In [39]:
# Limpiar la columna 'employeenumber'
df['employeenumber'] = df['employeenumber'].astype(str).str.replace(',', '').str.strip()  # Reemplazar comas y eliminar espacios en blanco

In [40]:
# Convertir la columna 'employeenumber' a tipo numérico (int), manejando los posibles errores
df['employeenumber'] = pd.to_numeric(df['employeenumber'], errors='coerce')

---------------------

No requiere cambios ✅

In [ ]:
df['EnvironmentSatisfaction'].unique()

array([ 1,  3,  4,  2, 42, 37, 35, 25, 27, 31, 39, 21, 15, 14, 33, 19, 12,
       13, 28, 47, 36, 29, 24, 46, 16, 22, 41, 49, 11, 48, 18, 10, 45, 38,
       17, 20, 26, 43])

-------------

- `cambiar de 0 - 1 a Male/Female o M/F`✅
- `convertir en object ya que antes era int`✅


In [ ]:
df['gender'] = df['gender'].replace({0: 'Male', 1: 'Female'})

In [ ]:
df['gender'].unique()

array(['Male', 'Female'], dtype=object)

--------

- `Cambiar a formato de hora  o int`✅
- `Eliminar el Not Available`✅

In [ ]:
# Rellenar los valores NA con 0 o cualquier otro valor predeterminado
df.hourlyrate = df.hourlyrate.replace({'Not Available': 0})


In [ ]:

# Convertir la columna a enteros
df['hourlyrate'] = df['hourlyrate'].astype(int)

In [ ]:
df['hourlyrate'].unique()

array([ 51,  65,  58,  82,  45,  99,  91,  64,  55,  68,  49,  61,  79,
        31,  69,  48,  80,  74,  98,  59,  33,  56,  66,  57,  53,  87,
        81,  84,  32,  41,  92,  47,   0,  43,  86,  30,  42,  88,  96,
        67,  62,  72,  78,  89,  52,  50,  90,  37,  94,  76,  60,  46,
        83, 100,  40,  97,  54,  75,  39,  85,  63,  44,  93,  36,  35,
        73,  71,  70,  38,  77,  95,  34])

------

No requiere cambios ✅

In [ ]:
df['jobinvolvement'].unique()

array([3, 2, 4, 1])

-------

No requiere cambios ✅

In [ ]:
df['joblevel'].unique()

array([5, 4, 3, 2, 1])

-----------

- `cambiar a minúsculas` ✅

In [ ]:
df['jobrole'] = df['jobrole'].str.lower()

In [ ]:
df['jobrole'].unique()

array([' research director ', ' manager ', ' sales executive ',
       ' manufacturing director ', ' research scientist ',
       ' healthcare representative ', ' laboratory technician ',
       ' sales representative ', ' human resources '], dtype=object)

-------

No requiere cambios ✅

In [ ]:
df['jobsatisfaction'].unique()

array([3, 4, 1, 2])

-------

- `cambiar a minúsculas` ✅
- `Eliminar los NAN`
- `Eliminar la plabras 'Marreid' ` ✅

In [ ]:
df["maritalstatus"] = df["maritalstatus"].str.lower()
df["maritalstatus"] = df["maritalstatus"].str.replace('marreid', 'married')

In [ ]:
df['maritalstatus'].unique()

array([nan, 'married', 'divorced', 'single'], dtype=object)

-------

- ` quitar las comas por puntos `✅
- `tiene NAN`

In [33]:
def cambiar_comas(cadena):
    try:
        return float(cadena.replace(',', '').replace('.', ''))
    except:
        return np.nan
df['monthlyincome'] = df['monthlyincome'].apply(cambiar_comas)

In [34]:
df['monthlyincome'].unique()

array([195370., 199990., 192320., 171690.,     nan, 171740., 165950.,
       199730., 134020., 132060., 195450., 180410., 192460., 107480.,
       167520.,  62010., 198450.,  40010., 104470., 160640.,  32100.,
       102660., 104750.,  61620.,  47210.,  46150., 169590., 103060.,
        54060.,  59020., 108550.,  59140.,  66460., 139730., 133200.,
        66870.,  47350., 138720., 190450., 160150.,  96130., 115100.,
        43060., 170460.,  50670.,  36920., 198470.,  23080.,  57470.,
       104220.,  63470.,  23480.,  30720., 124900.,  80200., 170680.,
        89430., 192720.,  55770.,  26910.,  74030.,  88230.,  35790.,
       177790., 182130., 135770., 191900., 171230., 191870., 100080.,
        79880.,  70830.,  47230.,  34070.,  29290., 120310., 154270.,
        51260.,  96190.,  50100., 190330., 104000.,  27930.,  56740.,
       191970.,  84120.,  33480.,  53090.,  76540., 128080.,  65860.,
        25190., 170990., 131200.,  31310.,  44440.,  66740.,  98520.,
       178610.,  744

-------

No requiere cambios ✅

In [ ]:
df['monthlyrate'].unique()

array([ 6462,  5678,  4933, ..., 15302, 26956, 16642])

-------------------------

No requiere cambios ✅

In [ ]:
df["numcompaniesworked"].unique()

array([7, 0, 1, 3, 2, 4, 8, 9, 5, 6])

-------

Eliminado ❌

In [ ]:
df['Over18'].unique()

-------

`hay un 42% de nulos, quizás eliminarla? - Eliminar los NAN`

In [ ]:
df['overtime'].isnull().sum()/df.shape[0]

0.4188351920693928

In [ ]:
df["overtime"].unique()

array(['No', nan, 'Yes'], dtype=object)

-------

No requiere cambios ✅

In [ ]:
df['percentsalaryhike'].unique()

array([13, 14, 11, 19, 12, 25, 16, 17, 22, 23, 20, 15, 21, 24, 18])

-------

- `volverlo int o float` ✅
- `Quitarle la coma y por puntos` ✅
- `Eliminar los NAN `

In [ ]:
df['performancerating'] = df['performancerating'].str.replace(",", ".")

In [ ]:
df['performancerating'] = df['performancerating'].astype(float)

In [ ]:
df["performancerating"].unique()

array([ 3.,  4., nan])

-------

No requiere cambios ✅

In [ ]:
df["relationshipsatisfaction"].unique()

array([3, 1, 4, 2])

-------

Eliminado ❌

In [ ]:
df["standardhours"].unique()

-------

No requiere cambios ✅

In [ ]:
df["stockoptionlevel"].unique()

array([0, 1, 2, 3])

-------

In [ ]:
# Funición unica que cambia los valores de las columnas de totalworkingyears, sameasmonthlyincome worklifebalance

def cambiar_comas_y_nulos(cadena):
    if pd.isnull(cadena):  # Si es un valor nulo, devuelve NaN
        return np.nan
    try:
        return float(cadena.replace(',', '.'))  # Intenta convertir la cadena a float
    except ValueError:  # Si la conversión falla, devuelve NaN
        return np.nan

# Lista de las columnas que deseas modificar
columnas_a_convertir = ['totalworkingyears', 'sameasmonthlyincome', 'worklifebalance']

# Aplica la función a cada columna en la lista
for columna in columnas_a_convertir:
    df[columna] = df[columna].apply(cambiar_comas_y_nulos)

- `Cambiar a int o float ` ✅
- `Eliminar los NAN `

In [ ]:
df["totalworkingyears"].unique()

array([nan, 34., 22., 28., 20., 21., 33., 40., 18., 25., 15., 17., 26.,
       16., 24., 14., 23., 27., 19., 11., 38., 37., 13., 12., 29., 10.,
       36., 35.,  9., 31., 32.,  8.,  7., 30.,  6.,  5.,  4.,  3.,  2.,
        1.,  0.])

-------

No requiere cambios ✅

In [ ]:
df["trainingtimeslastyear"].unique()

array([5, 3, 2, 0, 1, 4, 6])

-----------

- `Tiene NAN`
- `cambiar a int o float` ✅


In [ ]:
df['worklifebalance'].unique()

array([ 3., nan,  2.,  4.,  1.])

-------

No requiere cambios ✅

In [ ]:
df['yearsatcompany'].unique()

array([20, 33, 22, 19, 21, 18, 24, 31, 26, 16, 23, 15, 17, 32, 14, 13, 25,
       12, 11, 37, 40, 36, 27, 29, 10,  9, 30,  8,  7, 34,  6,  5,  4,  2,
        3,  1,  0])

-------

Eliminado ❌

- `tiene 97% null`

In [ ]:
df['yearsincurrentrole'].unique()

array([nan, '13,0', '12,0', '11,0', '7,0', '6,0', '4,0', '3,0', '2,0',
       '1,0', '0,0'], dtype=object)

-------

No requiere cambios ✅

In [ ]:
df['yearssincelastpromotion'].unique()

array([15, 11,  5,  2,  4,  7,  0,  1, 13, 14,  8, 12,  3,  6, 10,  9])

-------

No requiere cambios ✅

In [ ]:
df['yearswithcurrmanager'].unique() 

array([15,  9,  6,  8,  7, 11, 10, 12,  4,  0,  5, 17,  2, 14,  1, 13,  3,
       16])

-------

Eliminado ❌

In [ ]:
df["sameasmonthlyincome"].unique()

array([19537., 19999., 19232., 17169.,    nan, 17174., 16595., 19973.,
       13402., 13206., 19545., 18041., 19246., 10748., 16752.,  6201.,
       19845.,  4001., 10447., 16064.,  3210., 10266., 10475.,  6162.,
        4721.,  4615., 16959., 10306.,  5406.,  5902., 10855.,  5914.,
        6646., 13973., 13320.,  6687.,  4735., 13872., 19045., 16015.,
        9613., 11510.,  4306., 17046.,  5067.,  3692., 19847.,  2308.,
        5747., 10422.,  6347.,  2348.,  3072., 12490.,  8020., 17068.,
        8943., 19272.,  5577.,  2691.,  7403.,  8823.,  3579., 17779.,
       18213., 13577., 19190., 17123., 19187., 10008.,  7988.,  7083.,
        4723.,  3407.,  2929., 12031., 15427.,  5126.,  9619.,  5010.,
       19033., 10400.,  2793.,  5674., 19197.,  8412.,  3348.,  5309.,
        7654., 12808.,  6586.,  2519., 17099., 13120.,  3131.,  4444.,
        6674.,  9852., 17861.,  7446., 13757., 16598.,  4285., 10124.,
        9714.,  9924., 11245.,  3069.,  9241.,  6472.,  3055.,  8321.,
      

In [37]:
df.drop(columns=['sameasmonthlyincome'])

age attrition businesstravel  dailyrate  distancefromhome  education  \
0      51        No            NaN      684.0                 6          3   
1      52        No            NaN      699.0                 1          4   
2      42        No  travel rarely      532.0                 4          2   
3      47        No  travel rarely      359.0                 2          4   
4      46        No            NaN     1319.0                 3          3   
...   ...       ...            ...        ...               ...        ...   
1609   36       Yes  travel rarely      530.0                 3          1   
1610   45        No     non travel      805.0                 4          2   
1611   39        No  travel rarely      903.0               -13          5   
1612   36        No     non travel     1229.0                 8          4   
1613   46        No            NaN      566.0                 7          2   

        educationfield employeenumber  environmentsatisfaction  gender  \
0                  NaN          162,0                        1       0   
1        Life Sciences          259,0                        3       0   
2     Technical Degree          319,0                        3       0   
3              Medical            NaN                        1       1   
4     Technical Degree            NaN                        1       1   
...                ...            ...                      ...     ...   
1609     Life Sciences          967,0                        3       0   
1610               NaN          972,0                        3       0   
1611               NaN            NaN                       13       0   
1612  Technical Degree          990,0                        1       0   
1613           Medical            NaN                        4       0   

     hourlyrate  jobinvolvement  joblevel                   jobrole  \
0            51               3         5        resEArch DIREcToR    
1            65               2         5                  ManAGeR    
2            58               3         5                  ManaGER    
3            82               3         4        ReseArCH DIrECtOr    
4            45               4         4          sAleS EXECUtIve    
...         ...             ...       ...                       ...   
1609         51               2         3          saLEs ExeCUTiVe    
1610         57               3         2    LAboRaTOry tECHNiCIAn    
1611         41               4         3          sAlES ExECUTivE    
1612         84               3         2          SaLes ExecUtIVe    
1613         75               3         3   mAnUfactURInG DiRECTOr    

      jobsatisfaction maritalstatus  monthlyincome  monthlyrate  \
0                   3           NaN       195370.0         6462   
1                   3           NaN       199990.0         5678   
2                   4       Married       192320.0         4933   
3                   3       Married       171690.0        26703   
4                   1      Divorced            NaN         7739   
...               ...           ...            ...          ...   
1609                4       Married       103250.0         5518   
1610                2           NaN        44470.0        23163   
1611                3        Single            NaN         2560   
1612                4      Divorced            NaN        25952   
1613                3           NaN       108450.0        24208   

      numcompaniesworked overtime  percentsalaryhike performancerating  \
0                      7       No                 13               3,0   
1                      0      NaN                 14               3,0   
2                      1       No                 11               3,0   
3                      3      NaN                 19               3,0   
4                      2       No                 12               3,0   
...                  ...      ...                ...               ...   
1609

-------

`Es str, creo que no requiere cambios es una fecha de año por si sola`✅

In [ ]:
df["datebirth"].unique()

array([1972, 1971, 1981, 1976, 1977, 1975, 1964, 1982, 1967, 1985, 1968,
       1983, 1965, 1988, 1978, 1990, 1987, 1989, 1970, 1980, 1963, 1991,
       1986, 1974, 1984, 1973, 1979, 1993, 1994, 1992, 1969, 1966, 1996,
       1995, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005])

-------

Eliminado ❌

In [ ]:
df["salary"].unique()

-------

Eliminado ❌

In [ ]:
df['roledepartament'].unique()

------

Eliminado ❌

In [ ]:
df["numberchildren"].unique()

-------


`Modificamos los valores 0 y 1, True or False por Yes y No` ✅
- 1  =      360
- True  =   345
- 0   =     309
- False =   305
- Yes  =    295

In [ ]:
df['remotework']= df['remotework'].replace({('1'):'Yes',('0') or ('False'): 'No', 'True': 'Yes', 'False': 'No'})

In [ ]:
df["remotework"].unique()

array(['Yes', 'No'], dtype=object)

In [ ]:
df['remotework'].value_counts()

remotework
Yes    1000
No      614
Name: count, dtype: int64

-------

#### `Mover el employeenumber al inicio` ✅

In [ ]:
## Mover la columna 'employeenumber' al principio
df = df[['employeenumber'] + [col for col in df.columns if col != 'employeenumber']]

In [ ]:
# Guarda el DataFrame como un archivo CSV
df.to_csv("df_superlimpiooo.csv", index = False)

# Punto final --> Limpieza de Nulos

### `Variables categóricas con Nulos = Objeto`

In [ ]:
# Obtenemos la lista de columnas categóricas que tienen nulos
nulos_esta_cat = df[df.columns[df.isnull().any()]].select_dtypes(include = "O").columns
print("Las columnas categóricas que tienen nulos son : \n ")
print(nulos_esta_cat)

Las columnas categóricas que tienen nulos son : 
 
Index(['employeenumber', 'businesstravel', 'educationfield', 'maritalstatus',
       'overtime'],
      dtype='object')


In [ ]:
# sacamos el 'value_counts()' de cada una de las columnas categóricas que tienen nulos para saber como es la distribución de sus categorías
for col in nulos_esta_cat:
    print(f"La distribución de las categorías para la columna {col.upper()}")
    display(df[col].value_counts() / df.shape[0])  # display es una función utilizada para mostrar objetos de manera más legible en Jupyter Notebooks o entornos similares. 
    print("........................")

La distribución de las categorías para la columna EMPLOYEENUMBER


employeenumber
482,0     0.001239
530,0     0.001239
507,0     0.001239
517,0     0.001239
522,0     0.001239
            ...   
161,0     0.000620
164,0     0.000620
190,0     0.000620
194,0     0.000620
2040,0    0.000620
Name: count, Length: 1079, dtype: float64

........................
La distribución de las categorías para la columna BUSINESSTRAVEL


businesstravel
travel rarely        0.363073
travel frequently    0.102230
non travel           0.056382
Name: count, dtype: float64

........................
La distribución de las categorías para la columna EDUCATIONFIELD


educationfield
life sciences       0.216233
medical             0.171004
marketing           0.064436
technical degree    0.042751
other               0.036555
human resources     0.007435
Name: count, dtype: float64

........................
La distribución de las categorías para la columna MARITALSTATUS


maritalstatus
married     0.271995
single      0.201363
divorced    0.123296
Name: count, dtype: float64

........................
La distribución de las categorías para la columna OVERTIME


overtime
No     0.422553
Yes    0.158612
Name: count, dtype: float64

........................


### `Variables numéricas con Nulos = Int o float`


In [ ]:
# Obtenemos la lista de columnas numericas que tienen nulos
nulos_esta_num = df[df.columns[df.isnull().any()]].select_dtypes(include = np.number).columns
print("Las columnas numéricas que tienen nulos son : \n ")
print(nulos_esta_num)

Las columnas numéricas que tienen nulos son : 
 
Index(['dailyrate', 'monthlyincome', 'performancerating', 'totalworkingyears',
       'worklifebalance', 'sameasmonthlyincome'],
      dtype='object')


In [ ]:
# lo primero que vamos a hacer es calcular el número de nulos que tenemos en cada una de las columnas numericas
df[nulos_esta_num].isnull().sum() / df.shape[0]

dailyrate              0.076828
monthlyincome          0.522305
performancerating      0.120818
totalworkingyears      0.325898
worklifebalance        0.066914
sameasmonthlyincome    0.522305
dtype: float64

In [ ]:
# Guarda el DataFrame como un archivo CSV
df.to_csv("DF_Totalmente_Limpio.csv", index = False)